$NetBSD: patch-aa,v 1.2 2001/03/14 14:57:34 zuntum Exp $

IRIX: struct dirent has no d_type, so lstat() the entry; use strlen()
instead of the BSD-only d_namlen, and don't index before short names.

--- rtfm.c.orig
+++ rtfm.c
@@ -29,6 +29,7 @@
 
 #include <sys/types.h>
 #include <sys/param.h>
+#include <sys/stat.h>
 
 #include <ctype.h>
 #include <dirent.h>
@@ -51,25 +52,24 @@ extern char *__progname;
 
 static int cflag, dflag, iflag, mflag, tflag, wflag;
 
-/* From /usr/src/contrib/texinfo/info/filesys.h */
+/* From /usr/src/gnu/dist/texinfo/info/filesys.h */
 const char *info_paths[] = {
-	"/usr/local/info",
-	"/usr/info",
-	"/usr/local/lib/info",
+	"@LOCALBASE@/info",
+	"@LOCALBASE@/lib/info",
 	"/usr/lib/info",
-	"/usr/local/gnu/info",
-	"/usr/local/gnu/lib/info",
+	"@LOCALBASE@/gnu/info",
+	"@LOCALBASE@/gnu/lib/info",
 	"/usr/gnu/info",
 	"/usr/gnu/lib/info",
 	"/opt/gnu/info",
 	"/usr/share/info",
 	"/usr/share/lib/info",
-	"/usr/local/share/info",
-	"/usr/local/share/lib/info",
+	"@LOCALBASE@/share/info",
+	"@LOCALBASE@/share/lib/info",
 	"/usr/gnu/lib/emacs/info",
-	"/usr/local/gnu/lib/emacs/info",
-	"/usr/local/lib/emacs/info",
-	"/usr/local/emacs/info",
+	"@LOCALBASE@/gnu/lib/emacs/info",
+	"@LOCALBASE@/lib/emacs/info",
+	"@LOCALBASE@/emacs/info",
 	".",
 	NULL
 };
@@ -253,17 +253,26 @@ search_texinfo()
 		chdir(info_paths[i]);
 
 		while ((ent = readdir(dir)) != NULL) {
+#ifdef DT_REG
 			if (ent->d_type == DT_REG) {
+#else
+			/* No d_type (e.g. IRIX): lstat() it; we are in its dir. */
+			struct stat sb;
+
+			if (lstat(ent->d_name, &sb) == 0 && S_ISREG(sb.st_mode)) {
+#endif
 				char *command, c;
 				int matches = 0;
 				FILE *p;
+				/* d_namlen is not portable (Linux, IRIX). */
+				size_t namlen = strlen(ent->d_name);
 
 				/*
 				 * Anything that does not end in '.info.gz'
 				 * is not a Texinfo page.
 				 */
-				if (strcasecmp(".info.gz", &ent->d_name[ent->d_namlen - 8])
-				    && strcasecmp(".info", &ent->d_name[ent->d_namlen - 5]))
+				if ((namlen < 8 || strcasecmp(".info.gz", &ent->d_name[namlen - 8]))
+				    && (namlen < 5 || strcasecmp(".info", &ent->d_name[namlen - 5])))
 					continue;
 
 				/*
